core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -237,13 +237,17 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
         dynamicConfigManager.startup()
 
         /* tell everyone we are alive */
-        val listeners = config.advertisedListeners.map {case(protocol, endpoint) =>
-          if (endpoint.port == 0)
-            (protocol, EndPoint(endpoint.host, socketServer.boundPort(protocol), endpoint.protocolType))
-          else
-            (protocol, endpoint)
+        val advertisedListeners = config.advertisedListeners.mapValues { endpoint =>
+          val host =
+            if (Option(endpoint.host).map(_.trim).forall(host => host.isEmpty || host == "0.0.0.0"))
+              InetAddress.getLocalHost.getCanonicalHostName
```

Comment:
I am not sure using this right way to go here. InetAddress.getLocalHost() will ignore the resolv.conf and depend on the /etc/hosts to resolve the ip to hostname. In cases where users are depending on resolv.conf can experience issues. Why not we enforce users to add the hostnames in case of 0.0.0.0 as host.name